src/NamingStrategy/MapNamingStrategy.php

```diff
@@ -9,83 +9,100 @@
 
 namespace Zend\Hydrator\NamingStrategy;
 
-use Zend\Hydrator\Exception\InvalidArgumentException;
+use Zend\Hydrator\Exception;
 
 use function array_flip;
 use function array_key_exists;
 use function array_walk;
 use function is_int;
 use function is_string;
+use function sprintf;
 
-class MapNamingStrategy implements NamingStrategyInterface
+final class MapNamingStrategy implements NamingStrategyInterface
 {
     /**
-     * Map for hydrate name conversion.
-     *
      * @var string[]
      */
-    protected $mapping = [];
+    private $extractionMap = [];
 
     /**
-     * Reversed map for extract name conversion.
-     *
      * @var string[]
      */
-    protected $reverse = [];
+    private $hydrationMap = [];
 
     /**
-     * @param string[]      $mapping Map for name conversion on hydration
-     * @param null|string[] $reverse Reverse map for name conversion on extraction
+     * @param null|string[] $hydrationMap A map of string keys and values for
+     *     translation of hydrated field names. If not provided, the result of
+     *     an array_flip($extractionMap) will be used.
+     * @param null|string[] $extractionMap A map of string keys and values for
+     *     translation of extracted field names. If not provided, the result of
+     *     an array_flip($hydrationMap) will be used.
+     * @throws Exception\InvalidArgumentException if neither $hydrationMap nor
+     *     $extractionMap are provided.
+     * @throws Exception\InvalidArgumentException when flipping either the
+     *     $hydrationMap or $extractionMap, and any value is a non-string,
+     *     non-int value.
      */
-    public function __construct(array $mapping, ?array $reverse = null)
+    public function __construct(?array $hydrationMap = null, ?array $extractionMap = null)
     {
-        $this->mapping = $mapping;
-        $this->reverse = $reverse ?: $this->flipMapping($mapping);
+        if (null === $hydrationMap && null === $extractionMap) {
+            throw new Exception\InvalidArgumentException(sprintf(
+                '%s requires one or both of an array $hydrationMap and array $extractionMap;'
+                . ' neither provided',
+                __CLASS__
+            ));
+        }
+
+        if (null === $extractionMap && is_array($hydrationMap)) {
```

Comment:
`is_array` wasn't imported, but a not null check would also suffice here

Reply:
phpstan complained without the `is_array()` check; I actually added it in after-the-fact due to a warning phpstan raised.

Reply:
Resolved by using named constructors.